File: features/stage/src/main/java/com/nivasafinance/features/stage/controller/StageController.java
package com.nivasafinance.features.stage.controller;

import com.nivasafinance.common.constants.ApiConstants;
import com.nivasafinance.common.annotations.RequirePermission;
import com.nivasafinance.common.enums.EntityType;
import com.nivasafinance.common.utils.ValidationUtils;
import com.nivasafinance.features.rolemanagement.role.dto.UserAssignmentResponse;
import com.nivasafinance.features.rolemanagement.role.service.EntityOfficeKeyService;
import com.nivasafinance.features.stage.dto.StageFilterResponse;
import com.nivasafinance.features.stage.dto.StageKeysRequest;
import com.nivasafinance.features.stage.dto.StageTemplateResponse;
import com.nivasafinance.features.stage.service.StageReadService;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping(ApiConstants.V1 + "/stages")
@AllArgsConstructor
public class StageController {

    private final StageReadService stageReadService;
    private final EntityOfficeKeyService entityOfficeKeyService;

    @GetMapping
    @RequirePermission(permissionName = "READ_STAGE")
    public ResponseEntity<List<StageFilterResponse>> getAllActiveStages() {
        return ResponseEntity.ok(stageReadService.getAllActiveStages());
    }

    @PostMapping("/template")
    @RequirePermission(permissionName = "READ_STAGE")
    public ResponseEntity<Map<String, StageTemplateResponse>> getStageTemplates(
            @Valid @RequestBody StageKeysRequest request) {
        Map<String, StageTemplateResponse> templates = stageReadService.getStageTemplates(request.getStageKeys());
        return ResponseEntity.ok(templates != null ? templates : Collections.emptyMap());
    }

    
    @PostMapping("/assignable-users")
    @RequirePermission(permissionName = "READ_STAGE")
    public ResponseEntity<List<UserAssignmentResponse>> getAssignableUsersForStages(
            @Valid @RequestBody StageKeysRequest request,
            @RequestParam EntityType entityType,
            @RequestParam UUID entityId) {
        // Get entity's office to determine which users to show
        String officeKey = entityOfficeKeyService.getOfficeKey(entityType, entityId);
        if (!ValidationUtils.isNonNull(officeKey)) {
            return ResponseEntity.ok(Collections.emptyList());
        }
        
        List<UserAssignmentResponse> users = stageReadService.getAssignableUsersForStages(
                request.getStageKeys(), officeKey);
        return ResponseEntity.ok(users != null ? users : Collections.emptyList());
    }

    /**
     * Backward compatibility endpoint for getting assignable users for a single stage.
     * GET /api/v1/stages/{stageKey}/assignable-users
     */
    @GetMapping("/{stageKey}/assignable-users")
    @RequirePermission(permissionName = "READ_STAGE")
    public ResponseEntity<List<UserAssignmentResponse>> getAssignableUsersForStage(
            @PathVariable String stageKey,
            @RequestParam EntityType entityType,
            @RequestParam UUID entityId) {
        // Get entity's office to determine which users to show
        String officeKey = entityOfficeKeyService.getOfficeKey(entityType, entityId);
        if (!ValidationUtils.isNonNull(officeKey)) {
            return ResponseEntity.ok(Collections.emptyList());
        }
        
        List<UserAssignmentResponse> users = stageReadService.getAssignableUsersForStages(
                Collections.singletonList(stageKey), officeKey);
        return ResponseEntity.ok(users != null ? users : Collections.emptyList());
    }

    /**
     * Backward compatibility endpoint for getting template for a single stage.
     * GET /api/v1/stages/{stageKey}/template
     */
    @GetMapping("/{stageKey}/template")
    public ResponseEntity<StageTemplateResponse> getStageTemplate(
            @PathVariable String stageKey) {
        StageTemplateResponse template = stageReadService.getStageTemplate(stageKey);
        return ResponseEntity.ok(template);
    }
}

File: features/stages/src/main/java/com/nivasafinance/features/stages/controller/StageController.kt
package com.nivasafinance.features.stages.controller

import com.nivasafinance.features.stages.dto.StageRequest
import com.nivasafinance.features.stages.dto.StageResponse
import com.nivasafinance.features.stages.dto.StageUpdateRequest
import com.nivasafinance.features.stages.service.StageService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api/stages")
class StageController(
    private val stageService: StageService
) {

    @PostMapping
    fun createStage(
        @RequestBody stageRequest: StageRequest
    ): ResponseEntity<StageResponse> {
        val createdStage = stageService.createStage(stageRequest)
        return ResponseEntity.status(HttpStatus.CREATED).body(createdStage)
    }

    @GetMapping
    fun getAllStages(): ResponseEntity<List<StageResponse>> {
        val stages = stageService.getAllStages()
        return ResponseEntity.ok(stages)
    }

    @GetMapping("/{stageId}")
    fun getStageById(
        @PathVariable stageId: UUID
    ): ResponseEntity<StageResponse> {
        val stage = stageService.getStageById(stageId)
        return ResponseEntity.ok(stage)
    }

    @GetMapping("/definition/{stageDefinitionKey}")
    fun getStagesByDefinitionKey(
        @PathVariable stageDefinitionKey: String
    ): ResponseEntity<List<StageResponse>> {
        val stages = stageService.getStagesByDefinitionKey(stageDefinitionKey)
        return ResponseEntity.ok(stages)
    }

    @PutMapping("/{stageId}")
    fun updateStage(
        @PathVariable stageId: UUID,
        @RequestBody stageUpdateRequest: StageUpdateRequest
    ): ResponseEntity<StageResponse> {
        val updatedStage = stageService.updateStage(stageId, stageUpdateRequest)
        return ResponseEntity.ok(updatedStage)
    }
}
